File: 6-kyu/PatternCraft - State.js
class SiegeState {
  constructor() {  
  }
  canMove() { return false; }
  damage() { return 20; }
}

class TankState {
  constructor() {  
  }
  canMove() { return true; }
  damage() { return 5; }
}

class Tank {
  constructor() {
    this.state = new TankState();
  }
  get canMove() {return this.state.canMove(); }
  get damage() {return this.state.damage(); }
}

File: 6-kyu/PatternCraft - State.py
class TankState():
    def __init__(self):
        self.canMove = True
        self.damage = 5
        
class SiegeState():
    def __init__(self):
        self.canMove = False
        self.damage = 20
        
class Tank():
    def __init__(self):
        self.state = TankState()
    def can_move(self):
        return self.state.canMove
    def damage(self):
        return self.state.damage
